### DISCUSSION TITLE
ADL should be inhibited by default for non-operators

### DISCUSSION THREAD

**Discussion by KendallHarter** [hsutter/cppfront#781] [Suggestions]
Currently, the emitted function calls from cppfront are subject to ADL, which is a common pitfall in generic code, to the point that the standard ranges library is defined such that much of it does not participate in ADL.  By having some syntax that explicitly requests ADL it makes it much clearer when ADL is being intentionally used and makes generic code easier to write without needing to fully qualify almost every function call to have it be always correct.

**Comment by jcanizales** [comment]
I'm curious, why is ADL a pitfall in generic code?

**Reply by KendallHarter** [reply]
If you take in an object from another namespace any unqualified function call can result in a function call from that namespace, which often isn't what's wanted.  It's easy enough to suppress, either by qualifying the function name or placing it in parentheses.  I suppose the same pitfall exists with non-generic code, but generally it's easier to notice because you know the types that are being worked with.

**Reply by jcanizales** [reply]
> which often isn't what's wanted

But isn't the whole point of ADL for me to be able to define my own `min` or `swap` or whatever for my class, _so that_ your generic code doesn't have to care and will call the right one automatically?

Is the pitfall that the generic code would define `generic_library::f(T)` as some internal utility method, and call `f(t)` hoping to hit the former, instead going to `my_class::f(MyClass)`?

**Reply by JohelEGP** [reply]
Yes.
That's why C++ standard library implementations have to qualify most names.
Its most important provision regarding ADL is <https://eel.is/c++draft/global.functions#4>.

**Comment by JohelEGP** [comment]
For reference, here's the `adl` feature in Circle: <https://github.com/seanbaxter/circle/blob/master/new-circle/README.md#adl>.

A non-Cpp1 compiler tool like Cppfront would require you to qualify (almost) all unqualified function names with `adl`.

Maybe it'd be better to invest in a Clang-Tidy check that can be used from Cpp2.
The best I can think of without having to change Cpp2 is `/*adl*/` comments (if Clang-Tidy can work with that).
But Cppfront doesn't lower comments in the expected place (<https://cpp2.godbolt.org/z/rWzdvfE4f>):
```Cpp2
  /*adl*/ f(2);
  f(/*adl*/ g(2));
```
```C++
  f(2);/*adl*/
  f(        g(2));/*adl*/
```
